docs/index.rst

```diff
@@ -3,151 +3,713 @@
+resumable workflow.
+
+We will reveal the problem that we will be solving much later. For now, we can
+start by creating a couple of SQLAlchemy schemas:
+
+    * One to attach to the workflow itself (a workflow will represent a student)
+    * and one where a single grade (a grade will be the grade of a single test)
+
+Note that the `Workflow` model below can store an element pointer. This pointer
+(found at `engine.state.elem_ptr`) indicates the object that is currently being
+processed and saving it is crucial so that the engine can resume at a later
+time from that point.
 
 .. code-block:: python
 
-    def check_duplicate(obj, eng):
-       """
-       This task checks if the uploaded fulltext is a duplicate
-            @type obj: InspireGeneralForm
-            @precondition: obj.paths[]
-                    list, list of paths to uploaded files
-            @postcondition: obj.fulltext[]
-                    list containing txt for the extracted document
-                            obj.duplicateids[]
-                    list of inspire ids records that contain the duplicate of this document
-            @raise: stopProcessing on error
-            @return: True if duplicate found
-
-       """
-       ...
-
-So using the python docs, we can instruct workflow engine what types of
-arguments are acceptable, what is the expected outcome and what happens
-after the task finished.  And let's say, there will be a testing framework
-which will run the workflow pipeline with fake arguments and will test all
-sorts of conditions. So, the configuration is not cluttered with states
-and transitions that are possible, developers can focus on implementation
-of the individual tasks, and site admins should have a good understanding
-what the task is supposed to do -- the description of the task will be
-displayed through the web GUI.
-
-Here are some examples of workflow patterns (images are from
-http://www.yawlfoundation.org) and their implementation in
-Python. This gives an idea that workflow engine remains very simple
-and by supplying special functions, we can implement different
-patterns.
+    from sqlalchemy.ext.declarative import declarative_base
+    from sqlalchemy import Column, Integer, String, create_engine, ForeignKey, Boolean
+    from sqlalchemy.orm import sessionmaker, relationship
+
+    # Create an engine and a session
+    engine = create_engine('sqlite://')
+    Base = declarative_base(bind=engine)
+    DBSession = sessionmaker(bind=engine)
+    session = DBSession()
+
+
+    class Workflow(Base):
+        __tablename__ = 'workflow'
+        id = Column(Integer, primary_key=True)
+        state_elem_ptr = Column(Integer, default=-1)
+        grades = relationship('Grade', backref='workflow',
+                               cascade="all, delete, delete-orphan")
+
+        def save(self, elem_ptr):
+            """Save object to persistent storage."""
+            self.state_elem_ptr = elem_ptr
+            session.add(self)
```

Comment:
Indeed. If you meant `session.begin(subtransactions=True)`, then I changed it to use that.